quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -160,671 +164,714 @@ fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
-            let new = simplify_infix(&ac, InfixOperator::Plus, &bc, limit.saturating_sub(1));
-            min_by_key(original, new, size)
-        }
-
-        //----------------------------------------------------------------
-        // Finally: other parenthesis manipulation
-        //----------------------------------------------------------------
-
-        // Mul inside Div on left, multiplicand = denominator
-        (
-            Expression::Infix(InfixExpression {
-                left: ref multiplier,
-                operator: InfixOperator::Star,
-                right: ref multiplicand,
-            }),
-            InfixOperator::Slash,
-            ref denominator,
-        ) if **multiplicand == *denominator => *multiplier.clone(),
-
-        // Mul inside Div on left
-        (
-            ref numerator @ Expression::Infix(InfixExpression {
-                left: ref multiplier,
-                operator: InfixOperator::Star,
-                right: ref multiplicand,
-            }),
-            InfixOperator::Slash,
-            ref denominator,
-        ) => {
-            let original = div!(numerator.clone(), denominator.clone());
-            let new_multiplicand = simplify_infix(
-                multiplicand,
-                InfixOperator::Slash,
-                denominator,
-                limit.saturating_sub(1),
-            );
-            let new = simplify_infix(
-                multiplier,
+            // Addition associative, right: a + (b + c) = (a + b) + c, pick the shorter
+            (
+                ref a,
+                InfixOperator::Plus,
+                ref right @ Expression::Infix(InfixExpression {
+                    left: ref b,
+                    operator: InfixOperator::Plus,
+                    right: ref c,
+                }),
+            ) => {
+                let original = add!(a.clone(), right.clone());
+                let new_ab = simplify_infix(a, InfixOperator::Plus, b, limit - 1);
+                let new = simplify_infix(&new_ab, InfixOperator::Plus, c, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Addition associative, left: (a + b) + c = a + (b + c), pick the shorter
+            (
+                ref left @ Expression::Infix(InfixExpression {
+                    left: ref a,
+                    operator: InfixOperator::Plus,
+                    right: ref b,
+                }),
+                InfixOperator::Plus,
+                ref c,
+            ) => {
+                let original = mul!(left.clone(), c.clone());
+                let bc = simplify_infix(b, InfixOperator::Plus, c, limit - 1);
+                let new = simplify_infix(a, InfixOperator::Plus, &bc, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Multiplication associative, right: a * (b * c) = (a * b) * c, pick the shorter
```

Comment:
Could merge with the next case

Reply:
Same comment as before re: reconstructing the original